### PR TITLE
use composite buffers where possible

### PR DESCRIPTION
**Pull request by Outfluencer**
i think this should reduce memory allocations
I have tested it, its running without any issues for me.
But i did no perfromace tests for now, so they are welcome.

### CODE REVIEW

**Comment by Outfluencer** [comment]
oh you're probably right i tested it on an arm server (custom natives are not supported)

**Comment by Outfluencer** [comment]
but the FrameEncoder part should work right?

**Comment by Outfluencer** [comment]
> The naming of `fast` boolean in my opinion should be something like `canUseComposite`. And your complete removal from the length field prepender creates simplicity, but I think we could have it on fast and then if the CipherEncoder is added, we could set the prepender to's `canUseComposite` to false if we use natives, same for when adding compress.
> 
> https://github.com/SpigotMC/BungeeCord/blob/84ac7ab944dc911c333b4ef7d16e185c4c2625b7/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java#L510
> 
> 
> https://github.com/SpigotMC/BungeeCord/blob/master/proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java#L172
> (i would just leave out the logic to re-enable canUseComposite on compress removal if cipher not present tho)

sounds logic and simple, i will change.

**Comment by Outfluencer** [comment]
> (i would just leave out the logic to re-enable canUseComposite on compress removal if cipher not present tho)

why isn't it simple to just set the useComposite to true again?

**Comment by Janmm14** [comment]
> > The naming of `fast` boolean in my opinion should be something like `canUseComposite`. And your complete removal from the length field prepender creates simplicity, but I think we could have it on fast and then if the CipherEncoder is added, we could set the prepender to's `canUseComposite` to false if we use natives, same for when adding compress.
> 
> The prepender is an instance (its shared), should we create a new one per connection, or should we add another class and switch them in the pipeline. I prefer per channel Varint21LengthFieldPrepender

Yeah I would un-share it (change to per-connection) instead of replacing the impl.


> > (i would just leave out the logic to re-enable canUseComposite on compress removal if cipher not present tho)
> 
> why isn't it simple to just set useComposite to true again?

I thought that, as the minecraft protocol does not support this and when it was supported it could cause problems due to missing acknowledgement packet (as the threshold applies both ways), we could just not care about this potential performance penalty when compression is removed.

**Comment by Outfluencer** [comment]
I don't think that it could break anything. As there are no raceconditions because evey operation is performed on the same thread right?

**Comment by Janmm14** [comment]
> > The naming of `fast` boolean in my opinion should be something like `canUseComposite`. And your complete removal from the length field prepender creates simplicity, but I think we could have it on fast and then if the CipherEncoder is added, we could set the prepender to's `canUseComposite` to false if we use natives, same for when adding compress.
> 
> The prepender is an instance (its shared), should we create a new one per connection, or should we add another class and switch them in the pipeline. I prefer per channel Varint21LengthFieldPrepender

Oh yeah then the prepender should be per-channel.

> > (i would just leave out the logic to re-enable canUseComposite on compress removal if cipher not present tho)
> 
> why isn't it simple to just set useComposite to true again?

Well its possible, doesn't care, but as its a code path that's essentially unused

File: proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java [hunk 1]
```diff
@@ -187,5 +186,37 @@ public void setCompressionThreshold(int compressionThreshold)
...
+    /*
+     * Should be called on encryption add and on compressor add or remove
+     */
+    public void updateComposite()
+    {
+        CipherEncoder cipherEncoder = ch.pipeline().get( CipherEncoder.class );
+        PacketCompressor packetCompressor = ch.pipeline().get( PacketCompressor.class );
+        Varint21LengthFieldPrepender prepender = ch.pipeline().get( Varint21LengthFieldPrepender.class );
+        boolean compressorCompose = cipherEncoder == null || cipherEncoder.getCipher().allowComposite();
+        boolean frameDecoderCompose = compressorCompose && ( packetCompressor == null || packetCompressor.getZlib().allowComposite() );
```
Comment (Janmm14) [inline proxy/src/main/java/net/md_5/bungee/netty/ChannelWrapper.java:202]: It is an encoder, no decoder

**Comment by Janmm14** [comment]
purely codewise looks good to me now!

File: protocol/src/main/java/net/md_5/bungee/protocol/Varint21LengthFieldPrepender.java [hunk 2]
```diff
@@ -1,27 +1,38 @@
...
         int headerLen = varintSize( bodyLen );
-        out.ensureWritable( headerLen + bodyLen );
-
-        DefinedPacket.writeVarInt( bodyLen, out );
-        out.writeBytes( msg );
+        if ( compose )
+        {
+            ByteBuf buf = ctx.alloc().directBuffer( headerLen );
+            DefinedPacket.writeVarInt( bodyLen, buf );
+            list.add( ctx.alloc().compositeBuffer( 2 ).addComponents( true, buf, msg.retain() ) );
```
Comment (md-5) [inline protocol/src/main/java/net/md_5/bungee/protocol/Varint21LengthFieldPrepender.java:28]: Is there any reason why this isn't a direct composite but the compressor is?

Comment (Outfluencer) [reply protocol/src/main/java/net/md_5/bungee/protocol/Varint21LengthFieldPrepender.java:28]: By default it should allocate it direct but i can just call directCompositeBuffer method.
I will change it

**Review by md-5 — APPROVED** [review]

**Comment by andrewkm** [comment]
@md_5 this commit is causing all sorts of issues with various plugins, as has been reported to me by a few other server owners, not to mention it broke floodgate as well, thus I can't even properly test myself.

**Comment by md-5** [comment]
If it's a plugin not using the API then it's not really a bungee issue. Given that networking isn't exposed to plugins this is almost certainly the case.

**Comment by Outfluencer** [comment]
> @md-5 this commit is causing all sorts of issues with various plugins, as has been reported to me by a few other server owners, not to mention it broke floodgate as well, thus I can't even properly test myself.

Floodgate accessed a private static final field that does not exist anymore.

**Comment by Outfluencer** [comment]
@andrewkm i fixed the floodgate issue

**Comment by andrewkm** [comment]
> @andrewkm i fixed the floodgate issue

Thank you! :)

**Pull request closed without merging**